### ISSUE TITLE
RFC 20: need a way to express time limit in R?

### ISSUE THREAD

**Issue by garlick** [flux-framework/rfc#160]
Should there be a way to express an allocation time limit in _R_?

Practically speaking, some part of the system needs to throw a `timelimit` exception on a job once it's outlived its wallclock limit.   AFAICS the only thing we have to go on at this point is `duration` from jobspec, which is optional I think?

Another way to do it is when the scheduler places _R_ in the KVS before responding to a job manager allocation request, it could also write a wallclock expiration time to  another key like `job.active.<jobid>.R_expiration`.

**Comment by dongahn** [comment]
> Should there be a way to express an allocation time limit in _R_?

Yes, I do think we need it. 

> Practically speaking, some part of the system needs to throw a `timelimit` exception on a job once it's outlived its wallclock limit. AFAICS the only thing we have to go on at this point is `duration` from jobspec, which is optional I think?

Yes. The walltime should be determined as a function of the `duration` key in the jobspec and system-imposed policy. Generally speaking, we will need a way to create a system configuration to specify system imposed policy parameters. We had a related discussion at https://github.com/flux-framework/rfc/issues/101#issuecomment-329020512. Though the discussion was on the `exclusive` key, this really applies to a broader topic. 

> Another way to do it is when the scheduler places _R_ in the KVS before responding to a job manager allocation request, it could also write a wall-clock expiration time to another key like `job.active.<jobid>.R_expiration`.

We can add this to the `R_lite` key. But as you suggested, I like this separate key approach better since we will ultimately go to an API approach  and at that point `R_lite` will go away. @grondo?

**Comment by SteVwonder** [comment]
> AFAICS the only thing we have to go on at this point is `duration` from jobspec, which is optional I think?

I think there is also no guarantee that the `duration` in jobspec is a single integer.  It could be a range for some really advanced schedulers.

Having a `timelimit` directly in R could be really helpful when/if Flux instances are temporarily loaning resources to one another.  Then they would only have to send over a single payload with an R.

**Comment by dongahn** [comment]
> Generally speaking, we will need a way to create a system configuration to specify system imposed policy parameters.

For lack of the full system config support, this was my stopgap on flux-sched.

https://github.com/flux-framework/flux-sched/blob/master/resource/config/system_defaults.hpp#L28

**Comment by garlick** [comment]
I think configuration is a separate issue (not that it isn't an important one).  We could always use a broker attribute for the default time limit in the mean time.

Another thought is that maybe the time limit should be optional, if we want to support "cooperatively scheduled" work, where jobs get to run until they are done.  Such a mode that might be useful in a single user instance with a FCFS scheduler.

**Comment by dongahn** [comment]
> Another thought is that maybe the time limit should be optional, if we want to support "cooperatively scheduled" work, where jobs get to run until they are done. Such a mode might be useful in a single user instance with a FCFS scheduler.

The continuum model job within the pilot 2 workload would be an example. It needs to run until the entire allocation is done.

In this case, maybe the optional `duration` key shouldn't be specified so the instance scheduler will apply its max-time policy, which would be the time limit set for the entire instance?

**Comment by grondo** [comment]
> We can add this to the R_lite key. But as you suggested, I like this separate key approach better since we will ultimately go to an API approach and at that point R_lite will go away. @grondo?

I'm actually on the fence. Now that this has been brought up (thanks @garlick!), I think it would be handy if there was both a "creation time" and "expiration time" in `R`. In our current R v1 spec, this could be added under `execution` key as `starttime` and `duration` or perhaps better `expiration`.

It would be nice to have this information all in once place. Also, this would allow easier grow/shrink where each resource shard handed out to a job could have its own expiration time. Ideally you could trace the ctime of these R shards to see when exactly the job grew, and look at the R returned by the execution system to view when the job shrank.

On the other hand, the global job timelimit (even applied to a malleable job) is perhaps outside of any R, and it would be simpler to handle this as a separate key.

**Comment by garlick** [comment]
I like the idea of a separate time limit on resources and jobs!

Also, if we had an (optional?)  "starttime" in _R_, we could support the scheduler answering an allocation request with _R_ that begins in the future as a form of reservation.  (e.g. reserving viz resources for a demo at a scheduled meeting)

**Comment by dongahn** [comment]
> Also, if we had an (optional?) "starttime" in _R_, we could support the scheduler answering an allocation request with _R_ that begins in the future as a form of reservation. (e.g. reserving viz resources for a demo at a scheduled meeting)

If this were to be supported, "starttime" almost has to be mandatory? If you have two exactly same copies of resource set over two different points in time, how should the execution system release a specific copy of `R`?  An alternative would be to introduce some sort of an unique identifier for each `R`...

**Comment by dongahn** [comment]
> I like the idea of a separate time limit on resources and jobs!
I agree!

> I'm actually on the fence. Now that this has been brought up (thanks @garlick!), I think it would be handy if there was both a "creation time" and "expiration time" in `R`. In our current R v1 spec, this could be added under `execution` key as `starttime` and `duration` or perhaps better `expiration`.

I am fine with that. 

> Also, this would allow easier grow/shrink where each resource shard handed out to a job could have its own expiration time. Ideally you could trace the ctime of these R shards to see when exactly the job grew, and look at the R returned by the execution system to view when the job shrank.

This sounds good. But we will probably want to flesh out the detail for `grow`/`shrink` towards a fuller design once the current sprint is done.

**Comment by dongahn** [comment]
Oh. This remains me: do we need a standard for specifying a time or time quantity as well? I vaguely remember we discussed using an ISO standard, but I'm not sure if we ever reached a conclusion.

**Comment by garlick** [comment]
> do we need a standard for specifying a time or time quantity as well?

FWIW in [RFC 18](https://github.com/flux-framework/rfc/blob/master/spec_18.adoc#event-log-format) we use

> `timestamp` SHALL be the number of seconds since the Unix Epoch (1970-01-01 UTC), expressed with optional sub-second precision in decimal notation. It MUST be greater than zero.

**Comment by grondo** [comment]
> f you have two exactly same copies of resource set over two different points in time how should the execution system release a specific copy of R? 

The execution system will need to be able to release any valid *R* that is a subset of the resources currently allocated to it. It shouldn't have to relate to an existing *R* with a different starttime since currently allocated *R* can't overlap, unless I'm confused.

**Comment by dongahn** [comment]
Ah, true it probably isn't the execution system. 

The point I was trying to make was if we support "reserved allocation" and some core component want to release it (or a part of it), we may need a unique identifier or similar. Now that I think about it, maybe the job should suffice.

**Comment by dongahn** [comment]
@garlick and @grondo: want me to put together an PR with the below proposal and a time specificiation?

> I'm actually on the fence. Now that this has been brought up (thanks @garlick!), I think it would be handy if there was both a "creation time" and "expiration time" in `R`. In our current R v1 spec, this could be added under `execution` key as `starttime` and `duration` or perhaps better `expiration`.

I am currently working on the producer side so it would be better to add something firm quick.

**Comment by garlick** [comment]
> want me to put together an PR with the below proposal and a time specificiation?

Sure, thanks!

**Comment by garlick** [comment]
Closed by pr #164
